File: onebusaway-android/src/main/java/org/onebusaway/android/nav/NavigationReceiver.java
/*
 * Copyright (C) 2005-2019 University of South Florida
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.nav;

import android.app.NotificationManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import org.onebusaway.android.app.Application;

/**
 * Receives broadcasts when the user interacts with the navigation notification and passes them
 * to the NavigationServiceProvider
 */
public class NavigationReceiver extends BroadcastReceiver {
    public static final String TAG = "NavigationReceiver";

    public static final String NAV_ID = ".NAV_ID";
    public static final String ACTION_NUM = ".ACTION_NUM";
    public static final String NOTIFICATION_ID = ".NOTIFICATION_ID";

    public static final int DISMISS_NOTIFICATION = 1;
    public static final int CANCEL_TRIP = 2;

    @Override
    public void onReceive(Context context, Intent intent) {
        int navId = intent.getIntExtra(NAV_ID, 0);
        int actionNum = intent.getIntExtra(ACTION_NUM, 0);
        int notifyId = intent.getIntExtra(NOTIFICATION_ID, NavigationServiceProvider.NOTIFICATION_ID);

        switch (actionNum) {
            case DISMISS_NOTIFICATION:
                NavigationServiceProvider.mTTS.stop();
                break;

            case CANCEL_TRIP:
                cancelTrip(navId);
                break;
        }

    }

    private void cancelTrip(int navId) {
        Context appCxt = Application.get().getApplicationContext();
        appCxt.stopService(new Intent(appCxt, NavigationService.class));
        NotificationManager manager = (NotificationManager)
                appCxt.getSystemService(Context.NOTIFICATION_SERVICE);
        manager.cancel(NavigationServiceProvider.NOTIFICATION_ID);
    }

}

File: onebusaway-android/src/main/java/org/onebusaway/android/nav/NavigationReceiver.kt
/*
 * Copyright (C) 2005-2026 University of South Florida and Open Transit Software Foundation
 * Licensed under the Apache License, Version 2.0
 */
package org.onebusaway.android.nav

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.core.content.ContextCompat

/** Handles explicit notification commands without reaching into static service or TTS state. */
class NavigationReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        // Commands normally arrive as an action. The extra-based form is the shape used by
        // PendingIntents a previous app version created, which the system still holds and can
        // deliver here after an upgrade.
        val command = intent.action ?: when (intent.getIntExtra(ACTION_NUM, 0)) {
            DISMISS_NOTIFICATION -> ACTION_SILENCE
            CANCEL_TRIP -> ACTION_CANCEL
            else -> null
        }
        when (command) {
            ACTION_SILENCE -> ContextCompat.startForegroundService(
                context,
                Intent(context, NavigationService::class.java).setAction(NavigationService.ACTION_SILENCE)
            )
            ACTION_CANCEL -> ContextCompat.startForegroundService(
                context,
                Intent(context, NavigationService::class.java).setAction(NavigationService.ACTION_CANCEL)
            )
        }
    }

    companion object {
        // Extras and codes of the pre-modernization notification intents, kept only so upgraded
        // installs can still honour a PendingIntent the system created before the upgrade.
        const val ACTION_NUM = ".ACTION_NUM"
        const val DISMISS_NOTIFICATION = 1
        const val CANCEL_TRIP = 2

        const val ACTION_SILENCE = NavigationService.ACTION_SILENCE
        const val ACTION_CANCEL = NavigationService.ACTION_CANCEL
    }
}
